otaclient/app/ota_client.py

```diff
@@ -597,18 +563,123 @@ def rollback(self):
             failure_reason=self.last_failure_reason,
             failure_traceback=self.last_failure_traceback,
         )
-        if _live_ota_status == wrapper.StatusOta.UPDATING and self._update_executor:
-            _res.update_status = self._update_executor.get_update_status()
+        if live_ota_status == wrapper.StatusOta.UPDATING and self._update_executor:
+            status_report.update_status = self._update_executor.get_update_status()
+        return status_report
 
-        return _res
+
+class OTAClientBusy(Exception):
+    """Raised when otaclient receive another request when doing update/rollback."""
+
+
+class OTAClientWrapper:
+    """OTAClient stub implementation that wraps OTAClient, manage update/rollback session,
+    and exposes async API for OTAClientServiceStub.
+
+    All actual API request handlings are wrapped and dispatched into threadpool for execution.
+    Only one ongoing update/rollback is allowed, with the help of exclusive update_rollback lock.
+
+    If this ECU enables otaproxy and has child ECUs depend on it, control_flags is used to prevent
+    this ECU's reboot before the otaproxy dependency is resolved.
+
+    OTAClient itself is self-contained, its error handling is done internally, no exception will be raised
+    from calling API methods of OTAClient.
+    """
+
+    def __init__(
+        self,
+        *,
+        ecu_info: ECUInfo,
+        executor: ThreadPoolExecutor,
+        control_flags: OTAClientControlFlags,
+        proxy: Optional[str] = None,
+    ) -> None:
+        # only one update/rollback is allowed at a time
+        self.update_rollback_lock = asyncio.Lock()
+        self.my_ecu_id = ecu_info.ecu_id
+        self._run_in_executor = partial(
+            asyncio.get_running_loop().run_in_executor, executor
+        )
+
+        # create otaclient instance and control flags
+        self._otaclient = OTAClient(
+            boot_control_cls=get_boot_controller(ecu_info.get_bootloader()),
+            create_standby_cls=get_standby_slot_creator(cfg.STANDBY_CREATION_MODE),
+            my_ecu_id=self.my_ecu_id,
+            control_flags=control_flags,
+            proxy=proxy,
+        )
+
+        # proxy used by local otaclient
+        # NOTE: it can be an upper otaproxy, local otaproxy, or no proxy
+        self.local_used_proxy_url = proxy
+        self.last_operation = None  # update/rollback/None
+
+    # property
+
+    @property
+    def is_busy(self) -> bool:
+        return self.update_rollback_lock.locked()
+
+    # API
+
+    async def dispatch_update(self, request: wrapper.UpdateRequestEcu):
+        """Dispatch OTA update execution to background thread.
+
+        Raises:
+            OTAClientBusy if otaclient is already executing update/rollback.
+        """
+        if self.is_busy:
+            raise OTAClientBusy(f"ongoing operation: {self.last_operation=}")
+
+        async def _update():
+            async with self.update_rollback_lock:
```

Comment:
@obi-t4 Yes you are right! I will fix it right now. When the if statement is evaluated and the update lock is not acquired, it should be acquired at the point.

Reply:
Updated and fixed in [9bbac02](https://github.com/tier4/ota-client/pull/212/commits/9bbac0230abf7109312fcfd60d1fa3eaa346397d).

Reply:
@Bodong-Yang 
Let me confirm, update/(rollback) api can be called from the multiple of threads right?
If so, this lock doesn't work since asyncio.lock is not thread safe.
I think we need to limit service number as `grpc.aio.server(maximum_concurrent_rpcs=1)`?

Reply:
@obi-t4 

> Let me confirm, update/(rollback) api can be called from the multiple of threads right?

No, `OTAClientWrapper`'s APIs are async API, will only be called by otaclient's main thread, which runs `OTAClientServiceStub` instance(handling the incoming OTA update request). 
The flow is as follow: `grpc aio servicer -> OTAClientServiceStub.update -> OTAClientWrapper.dispatch_update`, the whole flow is done within the same async event loop, in the otaclient's main thread.

Reply:
> I think we need to limit service number as grpc.aio.server(maximum_concurrent_rpcs=1)?

This is not needed. the execution of co-routine is serialized(being executed one by one). With the latest update, 
```python
        # in OTAClientWrapper.dispatch_update
        if self.update_rollback_lock.locked():
            raise OTAClientBusy(f"ongoing operation: {self.last_operation=}")
        else:  # immediately take the lock if not locked
            await self.update_rollback_lock.acquire()
            self.last_operation = wrapper.StatusOta.UPDATING
```
There will be no context switching from the line `self.update_rollback_lock.locked()` to `await self.update_rollback_lock.acquire()`.

Reply:
The whole otaclient grpc server is async, and `otaclient_stub` and `OTAClientWrapper` are also fully async. 
Only actual OTA logics in `OTAClient` module are not async, so `OTAClientWrapper` will dispatch the update to the threadpool to run.